src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,291 @@
+				ring.logger.WithField("ring", ring.prefix).Errorf("failed rebalancing: %v", err)
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			ring.logger.WithFields(ring.memberLogFields()).Infof("shutting down ring")
+			return
+		}
+		ring.logger.WithFields(ring.memberLogFields()).Errorf("failed watch: %v", err)
+	}
+}
+
+func (ring *Ring) rebalance() error {
+	var errs []error
+	for key, lockInfo := range ring.node.Locks {
+		if ring.Get(key).Id == ring.node.Id {
+			continue
+		}
+		if err := lockInfo.Lock.Unlock(lockInfo.Ctx); err != nil {
+			errs = append(errs, errors.EnsureStack(err))
+		}
+		delete(ring.node.Locks, key)
+		ring.logger.WithFields(ring.lockLogFields(key)).Info("rebalanced lock")
+	}
+	if errs != nil {
+		return errors.New(fmt.Sprintf("errors unlocking locks while rebalancing: %v", errs))
+	}
+	return nil
+}
+
+// insertMemberById is a convenience wrapper for insertMember that constructs the Member on behalf of the caller.
+func (ring *Ring) insertMemberById(id string) {
+	member := Member{
+		Id:   id,
+		Hash: ring.hashFn([]byte(id)),
+	}
+	ring.insertMember(member)
+}
+
+// insertMember updates Ring.members, and sorts Ring.members by the Member hash.
+// If the Member already exists, insertMember is a no-op. The member can be mocked for testing purposes.
+func (ring *Ring) insertMember(member Member) {
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= member.Hash
+	})
+	if index < len(ring.members) && ring.members[index].Hash == member.Hash {
+		return // member already exists.
+	}
+	ring.logger.WithFields(ring.memberLogFields()).Infof("adding member %s to ring", member.Id)
+	if index >= len(ring.members) {
+		ring.members = append(ring.members, member)
+		return
+	}
+	ring.members = append(ring.members[:index+1], ring.members[index:]...)
+	ring.members[index] = member
+}
+
+// removeMemberById is a convenience wrapper for removeMember that constructs the Member on behalf of the caller.
+func (ring *Ring) removeMemberById(id string) {
+	member := Member{
+		Id:   id,
+		Hash: ring.hashFn([]byte(id)),
+	}
+	ring.removeMember(member)
+}
+
+// removeMember removes the Member from Ring.members if it exists.
+func (ring *Ring) removeMember(member Member) {
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= member.Hash
+	})
+	if ring.members == nil || index >= len(ring.members) || ring.members[index].Id != member.Id {
+		return // member doesn't exist.
+	}
+	ring.logger.WithFields(ring.memberLogFields()).Infof("deleting member %s from ring", member.Id)
+	ring.members = append(ring.members[:index], ring.members[index+1:]...)
+}
+
+// etcdToMember converts an etcd key in a ring's collection to a node ID.
+func (ring *Ring) etcdToMember(key string) string {
+	return path.Base(key)
+}
+
+// investigate tagged logger.
```

Comment:
Probably should put a `TODO` here.

Reply:
Actually I ended up removing this comment